lib/zcaps.js

```diff
@@ -0,0 +1,215 @@
+/*!
+ * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+// FIXME: use `base58-universal` instead
+const base58 = require('bs58');
+const brZCapStorage = require('bedrock-zcap-storage');
+const database = require('bedrock-mongodb');
+const {CapabilityDelegation} = require('ocapld');
+const {EdvClient} = require('edv-client');
+const jsigs = require('jsonld-signatures');
+const {extendContextLoader, SECURITY_CONTEXT_V2_URL, sign, suites} = jsigs;
+const {Ed25519Signature2018} = suites;
+const {util: {BedrockError}} = require('bedrock');
+
+// load config defaults
+require('./config');
+
+const ALLOWED_ACTIONS = {
+  read: ['read'],
+  write: ['read', 'write'],
+  issue: ['sign'],
+  kak: ['deriveSecret'],
+  hmac: ['sign', 'verify']
+};
+
+// delegates zcaps to the given invoker/delegator for the given instance
+exports.delegateCapabilities = async ({
+  instance, controllerKey, invoker, delegator, zcapMap
+}) => {
+  const signer = await controllerKey.getAsymmetricKey(instance.keys.zcapKey);
+
+  // delegate each type in `zcapMap`
+  const results = {};
+  for(const type in zcapMap) {
+    const parent = zcapMap[type];
+    // delegate zcap
+    let delegated;
+    const zcap = {
+      '@context': SECURITY_CONTEXT_V2_URL,
+      // use 128-bit random multibase encoded value
+      id: `urn:zcap:${await EdvClient.generateId()}`,
+      parentCapability: parent.id,
+      invoker,
+      delegator,
+      // FIXME: ensure ocapld.js checks allowedActions when verifying
+      // delegation chains
+      allowedAction: ALLOWED_ACTIONS[type]
+    };
+    if(type === 'kak' || type === 'hmac') {
+      zcap.referenceId = `${instance.id}-${type}`;
+      zcap.invocationTarget = {
+        id: parent.id,
+        type: parent.type
+      };
+      // delegate via controller key directly since it is the `controller`
+      // of the `kak` and the `hmac`
+      // FIXME: consider making zcap key `did:key` the controller instead
+      delegated = await _delegate({zcap, signer: controllerKey});
+    } else {
+      zcap.referenceId = parent.referenceId;
+      zcap.invocationTarget = {...parent.invocationTarget};
+      // delegate via externally referenced zcap key
+      delegated = await _delegate({zcap, signer});
+    }
+    results[delegated.referenceId] = delegated;
+  }
+  return results;
+};
+
+// FIXME: use did-key lib instead
```

Comment:
does the library referenced exist?  If not, does it do much more than provide the document loader implemented here?